OIIInspector/ImageManager.py

```diff
@@ -0,0 +1,151 @@
+            self._stop_image_registry(tolerate_err=True)
+        with suppress(RuntimeError):
+            self._remove_image_registry(tolerate_err=True)
+
+    def _stop_image_registry(self, tolerate_err=False):
+        run_cmd(f"{self._container_platform} stop {self._base_container_name}_{self._port}", tolerate_err=tolerate_err)
+
+    def _remove_image_registry(self, tolerate_err=False):
+        run_cmd(f"{self._container_platform} rm {self._base_container_name}_{self._port}", tolerate_err=tolerate_err)
+
+    def _check_if_image_already_runs(self):
+        return False
+
+    def _pull_image(self):
+        """
+        Pull needs to be done separately, because run can be interrupted with timer later (while running podman run).
+        """
+        run_cmd(f"{self._container_platform} pull {self._image_address}")
+        pass
+
+    def _serve_index_registry(self):
+        """
+        Locally start podman/docker image service, which can be communicated with using gRPC queries.
+        Resolution of port conflicts is handled in this function.
+        :return: tuple containing port number of the running service and the running Popen object.
+        :rtype: (int, Popen)
+        :raises OIIInspectorError: if all tried ports are in use, or the command failed for another reason.
+        """
+        port_start = self._grpc_start_port
+        port_end = port_start + self._grpc_max_port_tries
+
+        for port in range(port_start, port_end):
+            try:
+                return (
+                    port,
+                    self._serve_index_registry_at_port(
+                        port, self._grpc_max_tries, self._grpc_init_wait_time),
+                )
+            except AddressAlreadyInUse:
+                log.info('Port %d is in use, trying another.', port)
+
+        err_msg = "No free port has been found after {number_of_ports} attempts.".format(
+            number_of_ports=self._grpc_max_port_tries)
+        log.error(err_msg)
+        raise Exception(err_msg)
+
+    @retry(exceptions=OIIInspectorError, tries=1, logger=log)
+    def _serve_index_registry_at_port(self, port, max_tries, wait_time):
+        """
+        Start an image registry service at a specified port.
+        :param str int port: port to start the service on.
+        :param max_tries: how many times to try to start the service before giving up.
+        :param wait_time: time to wait before checking if the service is initialized.
+        :return: object of the running Popen process.
+        :rtype: Popen
+        :raises OIIInspectorError: if the process has failed to initialize too many times, or an unexpected
+        error occured.
+        :raises AddressAlreadyInUse: if the specified port is already being used by another service.
+        """
+        cmd = f"{self._container_platform} run --name={self._base_container_name}_{port} " \
+              f"-p={port}:{port} {self._image_address}"
+        for attempt in range(max_tries):
```

Comment:
There's already a retry decorator, so using max tries as well is unnecessary (in fact it will be too many retries)